openeo/testing/results.py

```diff
@@ -88,6 +91,129 @@ def _as_xarray_dataarray(data: Union[str, Path, xarray.DataArray]) -> xarray.Dat
+
+    :return: list of issues (empty if no issues)
+    """
+    # TODO: make this a public function?
+    # TODO: option for nodata fill value?
+    # TODO: option to include data type check?
+    # TODO: option to cast to some data type (or even rescale) before comparison?
+    # TODO: also compare attributes of the DataArray?
+    actual = _as_xarray_dataarray(actual)
+    expected = _as_xarray_dataarray(expected)
+    issues = []
+
+    if actual.dims != expected.dims:
+        issues.append(f"Dimension mismatch: {actual.dims} != {expected.dims}")
+    for dim in sorted(set(expected.dims).intersection(actual.dims)):
+        acs = actual.coords[dim].values
+        ecs = expected.coords[dim].values
+        if not (acs.shape == ecs.shape and (acs == ecs).all()):
+            issues.append(f"Coordinates mismatch for dimension {dim!r}: {acs} != {ecs}")
+    if actual.shape != expected.shape:
+        issues.append(f"Shape mismatch: {actual.shape} != {expected.shape}")
+
+    if not issues:
+        threshold = abs(expected * rtol) + atol
+        diff_exact = abs(expected - actual)
+        diff_mask = diff_exact > threshold
+        diff_lenient = diff_exact.where(diff_mask)
+
+        non_x_y_dims = list(set(expected.dims) - {"x", "y"})
+        value_mapping = dict(map(lambda d: (d, expected[d].data), non_x_y_dims))
+        shape = tuple([len(value_mapping[x]) for x in non_x_y_dims])
+
+        for shape_index, v in np.ndenumerate(np.ndarray(shape)):
+            indexers = {}
+            for index, value_index in enumerate(shape_index):
+                indexers[non_x_y_dims[index]] = value_mapping[non_x_y_dims[index]][value_index]
+            diff_data = diff_lenient.sel(indexers=indexers)
+            total_pixel_count = expected.sel(indexers).count().item()
+            diff_pixel_count = diff_data.count().item()
+
+            if diff_pixel_count > 0:
+                diff_pixel_percentage = round(diff_pixel_count * 100 / total_pixel_count, 1)
+                diff_mean = round(diff_data.mean().item(), 2)
+                diff_var = round(diff_data.var().item(), 2)
+
+                key = ",".join([f"{k} {str(v1)}" for k, v1 in indexers.items()])
+                issues.append(
+                    f"{key}: value difference exceeds tolerance (rtol {rtol}, atol {atol}), min:{diff_data.min().data}, max: {diff_data.max().data}, mean: {diff_mean}, var: {diff_var}"
+                )
+
+                coord_grid = np.meshgrid(diff_data.coords["x"], diff_data.coords["y"])
+
+                mask = diff_data.notnull()
+                if mask.dims[0] != "y":
+                    mask = mask.transpose()
+                c1 = coord_grid[0][mask]
+                c2 = coord_grid[1][mask]
+                coordinates = np.dstack((c1, c2)).reshape(-1, 2)
+
+                art = ascii_art(diff_data)
+                print(f"Difference ascii art for {key}")
+                print(art)
```

Comment:
I also have my doubts about using a raw `print` here, because that might not play well with pytest based reporting (e.g. we also try to produce a HTML report in APEx).

I guess it's a matter of experimenting  on what currently happens here, or what alternatives are possible

Reply:
Ok it does seem to work in the HTML report:
![image](https://github.com/user-attachments/assets/149a5d06-58cb-4421-877d-c4e3789e67bf)

but as mentioned elsewhere, the ascii art is a bit too large for good overview in that context